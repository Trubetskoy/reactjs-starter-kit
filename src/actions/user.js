import API from '../services/api'
import sort from '../services/sort'
export const LOGIN = 'LOGIN'
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS'
export const LOGIN_FAILURE = 'LOGIN_FAILURE'

export function login (email, password) {
  return async dispatch => {
    dispatch({ type: LOGIN })
    try {
      let res = await API.post({ email, password })
      res.categories = sort(res.categories)
      dispatch({ type: LOGIN_SUCCESS, payload: res })
      return res
    } catch (err) {
      dispatch({ type: LOGIN_FAILURE })
    }
  }
}
