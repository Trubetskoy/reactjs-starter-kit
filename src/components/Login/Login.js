import React, { useCallback, useState } from 'react'
import PropTypes from 'prop-types'
import { isEmailValid, isPasswordValid} from '../../services/validation'

const defaultError = {status: false, message: ''}

const Login = ({login, name, isRequesting, isFailure, isLoggedIn}) => {
  const [email, setEmail] = useState('')
  const [emailError, setEmailError] = useState(defaultError)
  const [password, setPassword] = useState('')
  const [passwordError, setPasswordError] = useState(defaultError)

  const onSubmit = useCallback(() => {
    setEmailError(isEmailValid(email) ?  {...defaultError} : setEmailError({status: true, message: 'Email is not valid'}));
    setPasswordError(isPasswordValid(email) ? {...defaultError} : {status: true, message: 'Password is not valid'})
    if (!emailError.status && !passwordError.status) {
      login(email, password)
    }
  }, [email, password])

  return (
    <div className='login'>
      <div className='login-header'>
        <h2>Welcome to login</h2>
      </div>
      <p className='login-intro'>
        <code>src/components/Login/Login.js</code>
      </p>
      {!isLoggedIn && (
        <div className='login-form'>
          <div>
            <label htmlFor='email'>Email</label>
            <input id='email' type='text' name='email' value={email} onChange={(event) => setEmail(event.target.value)} />
          </div>
          <div>
            <label htmlFor='password'>Password</label>
            <input id='password' type='password' name='password' value={password} onChange={(event) => setPassword(event.target.value)} />
          </div>
          {isFailure && (<div className='error'>Something went wrong</div>)}
          <div>
            <button data-testid='submit' onClick={onSubmit}>Submit</button>
          </div>
        </div>
      )}
      {isLoggedIn && (
        <div data-testid='user-data' className='user-data'>
          <div>
            <label htmlFor='name'>Name</label>
            <div id='name'>{name}</div>
          </div>
          <div>
            <label htmlFor='email'>Email</label>
            <div id='email'>{email}</div>
          </div>
        </div>
      )}
      {isRequesting && (<div className='spinner'>Loading...</div>)}
    </div>
  )
}

Login.propTypes = {
  login: PropTypes.func,
  name: PropTypes.string,
  isRequesting: PropTypes.bool,
  isFailure: PropTypes.bool,
  isLoggedIn: PropTypes.bool
}

Login.defaultProps = {
  login: () => {},
  name: '',
  isRequesting: false,
  isFailure: false,
  isLoggedIn: false
}

export default Login
