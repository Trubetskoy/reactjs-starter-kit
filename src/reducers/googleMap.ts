import { CREATE_MAP_POINT, DELETE_MAP_POINT } from "../constants";
import update from "immutability-helper";

const defaultMapState: any = [];

export default (mapState = defaultMapState, action: any) => {
  switch (action.type) {
    case CREATE_MAP_POINT:
      console.log(action.payload)
      return update(mapState, { $push: [action.payload] });
    case DELETE_MAP_POINT:
      return {
        ...mapState
      };

    default:
      return mapState;
  }
};
