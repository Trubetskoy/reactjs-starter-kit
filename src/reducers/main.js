import {CHANGE_STATE_PROP} from '../actions'

const REDUCER = 'MAIN'
const defaultState = {
}

export default (state = defaultState, action) => {
  switch (action.type) {
    case REDUCER + CHANGE_STATE_PROP:
      state[action.state.prop] = action.state.value
      return {
        ...state
      }
    default:
      return state
  }
}
