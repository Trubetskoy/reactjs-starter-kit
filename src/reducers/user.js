import update from 'immutability-helper'
import { CHANGE_STATE_PROP } from '../actions'
import { LOGIN, LOGIN_FAILURE, LOGIN_SUCCESS } from '../actions/user'

const REDUCER = 'USER'
export const defaultState = {
  name: '',
  categories: [],
  isRequesting: false,
  isFailure: false,
  isLoggedIn: false
}

export default (state = defaultState, action) => {
  switch (action.type) {
    case LOGIN:
      return {
        ...state,
        isRequesting: true
      }
    case LOGIN_SUCCESS:
      return {
        ...state,
        ...action.payload,
        isRequesting: false,
        isLoggedIn: true
      }
    case LOGIN_FAILURE:
      return {
        ...defaultState,
        isRequesting: false,
        isFailure: true
      }
    case REDUCER + CHANGE_STATE_PROP:
      return update(state, {
        [action.state.prop]: {$set: action.state.value}
      })
    default:
      return state
  }
}
