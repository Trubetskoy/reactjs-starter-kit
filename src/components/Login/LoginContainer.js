import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import Login from './Login'
import { changeStateProp } from '../../actions'
import { login } from '../../actions/user'

const mapStateToProps = ({user: {name, email, isRequesting, isFailure, isLoggedIn}}, ownProps) => {
  return {
    name,
    email,
    isRequesting,
    isFailure,
    isLoggedIn
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    ...bindActionCreators({changeStateProp, login}, dispatch)
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Login)
