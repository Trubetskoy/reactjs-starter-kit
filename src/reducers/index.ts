import main from "./main";
import ui from "./ui"

export default { main, ui };
