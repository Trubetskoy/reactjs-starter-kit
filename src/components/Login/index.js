import Login from './Login'
import LoginContainer from './LoginContainer'

export {
  Login,
  LoginContainer
}

export default {
  Login,
  LoginContainer
}
