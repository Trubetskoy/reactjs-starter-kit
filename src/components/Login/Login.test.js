/* eslint-env jest */
import React from 'react'
import { render, getByLabelText, getByTestId, getByText, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import Login from './Login'
import { defaultState } from '../../reducers/user'

describe('<Login /> tests', () => {
  it('should render as default login form', () => {
    const {container} = render(<Login {...defaultState}/>)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render with class', () => {
    const {container} = render(<Login {...defaultState}/>)
    expect(container.firstChild).toHaveClass(`login`)
  })

  it('should two inputs and button', () => {
    const {container} = render(<Login {...defaultState}/>)
    const username = getByLabelText(container, 'Email')
    const password = getByLabelText(container, 'Password')
    const button = getByTestId(container, 'button')
    expect(username).toHaveAttribute(`name`, 'email')
    expect(username).toHaveAttribute(`id`, 'email')
    expect(password).toHaveAttribute(`name`, 'password')
    expect(password).toHaveAttribute(`id`, 'password')
    expect(button).toBeInTheDocument()
  })

  it('should call method on button click', () => {
    const submit = jest.fn()
    const {container} = render(<Login login={submit} />)
    const button = getByTestId(container, 'button')
    fireEvent.click(button)
    expect(submit).toHaveBeenCalledTimes(1)
  })

  it('should render props during login', () => {
    let props = {
      ...defaultState,
      isRequesting: true
    }
    const {container} = render(<Login {...props} />)
    const spinner = getByText(container, 'Loading...')
    expect(spinner).toBeInTheDocument()
  })

  it('should render props after login success', () => {
    let props = {
      ...defaultState,
      name: 'test',
      isLoggedIn: true
    }
    const {container} = render(<Login {...props} />)
    const section = getByTestId(container, 'user-data')
    expect(section).toBeInTheDocument()
    const name = getByText(container, 'test')
    expect(name).toBeInTheDocument()
  })

  it('should render props after login failure', () => {
    let props = {
      ...defaultState,
      isFailure: true
    }
    const {container} = render(<Login {...props} />)
    const error = getByText(container, 'Something went wrong')
    expect(error).toBeInTheDocument()
  })
})
