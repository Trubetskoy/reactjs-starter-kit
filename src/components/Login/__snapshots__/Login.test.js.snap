// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> tests should render as default login form 1`] = `
<div
  class="login"
>
  <div
    class="login-header"
  >
    <h2>
      Welcome to login
    </h2>
  </div>
  <p
    class="login-intro"
  >
    <code>
      src/components/Login/Login.js
    </code>
  </p>
  <div
    class="login-form"
  >
    <div>
      <label
        for="email"
      >
        Email
      </label>
      <input
        id="email"
        name="email"
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        for="password"
      >
        Password
      </label>
      <input
        id="password"
        name="password"
        type="password"
        value=""
      />
    </div>
    <div>
      <button
        class="button-blue"
        data-testid="button"
      >
        Login
      </button>
    </div>
  </div>
</div>
`;
